2025 Remake

endulum opened the pull request:
The Shuffle Clock has gone 2+ years without a proper facelift, new features, and the TypeScript rewrite it deserves.

There does exist a `react-ts-remake` branch, but it's old, a Vercel deployment (which had been shared with multiple other people) depends on it, and perhaps this deserves its own branch anyway.

From my other major solitary projects since then, I've learned a whole lot, and I want to see if I could apply fresher eyes to this project and its structure.

### Todo

- [x] Update linting
- [x] Add Prettier
- [x] Review Settings structure and behavior
- [x] Thorough rewrite of styling and components

### Major issues

- [ ] ~~Still cannot figure out how to keep time and send notifications when browser or window is no longer in focus on mobile.~~ No sufficient workarounds to mobile throttling found this time, project will be produced as is and there will be an issue opened about this afterward.

endulum committed:
working pausable clock, can set delay

endulum committed:
migrated assets, styling existing elements

endulum committed:
expanding accordions for sound and notifs

endulum committed:
Sound and Notifs are too specialized to combine, separated them

endulum committed:
tidying up App, made switch labels shorter

endulum committed:
minimized component amount with some combining

endulum committed:
some lint and a11y tweaks

endulum committed:
inputs for the rest of the settings

endulum committed:
completed alert handling, moved audio to public

endulum committed:
new linter?

endulum committed:
replace expand btn with test btn for sound and notif

endulum committed:
added hacktimer

endulum committed:
instanceOfSettings to check local storage data

endulum committed:
further styling adjusts

endulum committed:
footer and some other tweaks

endulum committed:
adjust linting config

endulum committed:
set everything aside, start with empty App and one clockSettings hook

endulum committed:
add and test helper for transforming stored settings

endulum committed:
flesh out useClockSettings hook

endulum committed:
add SettingsFields component and wire up inputs to settings

endulum committed:
add ToggleLabel component for settings that should have switches

endulum committed:
add specialized NotifToggleLabel for notification support and toggling

endulum committed:
add service worker and function for notifying, add function for sound play

endulum committed:
fix some errors for deployment

endulum committed:
disable notif test if not possible, tweak some styling

endulum committed:
add sound settings, custom sound input attempt

endulum committed:
add audio element to use custom sound url

endulum committed:
extract sound url field as component, accept urls as playable sounds

endulum committed:
add component for notifications settings group

endulum committed:
add svg and another condition to sound url error text

endulum committed:
allow sw notification to link to biome on click

endulum committed:
differentiate notif and sw support, prioritizing notifs

endulum committed:
attempt fix for notif support false positive

endulum committed:
undo that fix...

endulum committed:
further tweaking to notif support

endulum committed:
extract notif and sw function from doNotify

endulum committed:
fix unused notifSupport param in extracted notif functions

endulum committed:
further notif check tweaking

endulum committed:
revive clock component

endulum committed:
correct clock alert time, adjust playSound params, implement hacktimer

endulum committed:
extract clock time logic to hook, extract shuffle animation to function

endulum committed:
add DelayField component

endulum committed:
tidy up css

endulum committed:
set page title to show time left until shuffle if not paused

endulum committed:
fix missing style for .disabled class

endulum committed:
add footer content

endulum committed:
delete trash

endulum committed:
remove comments, set aria-hidden on audio element

endulum committed:
update testing for transformSettings helper

endulum committed:
remove useless comments

endulum committed:
extract notifText string and pass to notify functions

endulum committed:
log of alert times in session storage

endulum committed:
add modal that displays logs

endulum committed:
add function to (hopefully) intermittently wake sw

endulum committed:
reformat logs to have timestamp and message, log possible sound + notif errors

endulum committed:
move files, extract useNotifSettings hook, other fiddling

endulum committed:
move Delay settings to inside Settings, add noHourly and noDelayOnHourly to clock settings

endulum committed:
fix missing default checked states for noDelay and noHourly

endulum committed:
rename event log, use generated timestamps instead of params

endulum committed:
adjust min height of body

endulum committed:
ability to store uploaded sounds in state and play them

endulum committed:
create useCustomAudio hook, indicate need for Settings cleanup

endulum committed:
some cleanup to Settings, store uploaded sounds

endulum committed:
make footer icons open their links in new tab

endulum committed:
turn handleInput into a generic

endulum committed:
adjust biome select mapping

endulum committed:
change audioString let to const

endulum committed:
move toggle focus appearance to css instead of handlers

endulum committed:
try to have the sworker keep the page awake

endulum committed:
implement NoSleep

endulum committed:
pull noSleep out of condition, log whether user is on mobile or not

endulum committed:
extract SettingBody component to allow for hiding disabled settings

endulum committed:
fix disabled settings still able to be tabbed to

endulum committed:
update readme

endulum committed:
fix SettingBody pointer blocking being on wrong condition

endulum committed:
move css hack to a div hack for SettingBody cutting off one pixel

endulum committed:
add manifest.json for installability?

endulum committed:
add custom hourly delay as an option

endulum committed:
add and test getAlertString func, allow for negative hourly delays

endulum committed:
fix custom hourly text not greyed when noHourly is true

endulum committed:
suppress vercel deployments for this branch

endulum committed:
tidying

- replace legacy eslint config
- drop airbnb+standard configs, don't really want these, just use basic
  recs
- update packaging
- add some packages and a config to make vitest work
- add prettier
- inadverdent formatting when checking out and tweaking test files and
  the functions they were regarding

endulum committed:
underscoring

add underscore to gitignore and underscore most contents of src, to keep as reference

endulum committed:
draft top level components, populate assets and styling

endulum committed:
create clock

- install material icons
- flesh out clock component and hook

endulum committed:
invoke notifications through service worker

- flesh out service worker file with event listeners and notes on notif
  scheduling
- create module for service worker messages and callback wrapper that
  checks sw support
- give the `useClock` hook parameters for functions on pause, play, and
  alert

endulum committed:
add and style switch

endulum committed:
move settings functions and hook out of ignored dirs, tweaks

endulum committed:
implement context for settings toggles

- create context, use context to pass clock settings and setter
- have switch component access context to use setter
- have mgmt components access context to use settings

endulum committed:
give context more general name

endulum committed:
add notification permission and support to app context

- pull out `useNotifSettings` hook from ignored files
- move permission ask functionality and error catching for it into hook
- add note about checking notification support
- reformat notif mgmt component to display ask button/blocked status

endulum committed:
create and test function to get ms until next expected alert

endulum committed:
temporarily change priority of notif support to favor sworker

endulum committed:
ready worker and clock for dummy timeout

endulum committed:
import hacktimer

endulum committed:
unsuppress deployment

endulum committed:
ignore getAlertString file, have vite bundler ignore underscored files

endulum commented:
It's looking more and more difficult to get continuous background timekeeping working.

To start with, the public raw-JS deployment of the Shuffle Clock relies on the Notifications API and [HackTimer](https://github.com/ctubio/HackTimer) to stay awake to count time, and serve notifications. This is sufficient in desktop environments, keeping the clock perfectly operational for an indefinite amount of time, no matter whether the page is unfocused or not.

Both capabilities are hindered on mobile:
- Support for the Notification API is spotty, and invoking `Notification.create()` can throw an error, even if the check for the API's presence succeeds (`if ('Notification' in window)`).
- HackTimer does not stop the page from being throttled or put to sleep by the browser or device when the page or app is unfocused. I've learned that the mobile environment can get more aggressive about throttling, and has more reason to (battery is a huge one).

On the `react-ts-remake` branch, the Clock overcame the Notification shortcoming by falling back to a Service Worker to deliver notifications. So far, there hasn't been an issue invoking them as such, which leaves the other issue; preventing sleep:
- use of [nosleep.js](https://github.com/richtr/NoSleep.js) had no effect
- making the app PWA-compatible and having it installed had no effect
- intermittently posting an event to Service Worker to access console or localStorage to "force" the app page awakening had no effect (which should have been obvious to me as I was using the *app* to post events... the app, which is prone to *sleeping*...)

After all that, I'll ultimately assume that mobile throttling is aggressive enough that relying on in-app code to keep intervals and count time is futile.

Service Workers, though, operate independently of the app, so I wondered about having the Service Worker invoke notifications on a timer-based schedule, rather than having the app itself do the invocation. 

This could theoretically work indefinitely; however, when not being inspected by console, Service Workers are killed 30 seconds after not receiving an event, something I monitored myself and later confirmed with [this article](https://developer.chrome.com/blog/eyeos-journey-to-testing-mv3-service%20worker-suspension). I could have the app post events to the Service Worker, but this is of course futile once the app gets put to sleep. The only other source of events for a Service Worker is a server.

[Periodic Sync](https://developer.mozilla.org/en-US/docs/Web/API/PeriodicSyncManager) registration seemed promising to resolve this, except:
- You can only set a *minimum* sync interval - no guarantee of by-the-second accuracy the Clock requires - and it's up to the browser to choose a reasonable sync time, which sucks because...
- ... most browsers' reasonable sync interval timing is a few hours.
Either way, I've just never gotten it to work.

I'm a good few levels deep in Workaround Hell, and considering just giving up, short of:
- Developing a standalone mobile app. I don't have the time or willpower to learn to do that right now, and that's not even considering the possibility that by-the-second intervals could *also* be throttled in a standalone app.
- Developing a timekeeping push server. Server-subscribed Service Workers *do* wake up and respond to server events. I know how to make and deploy a server. Fitting a server to this specific purpose could take some time and willpower, and poses a risk of going down Workaround Hell once again. I also just... don't *want* to engineer one, for an app meant to be simple.

I may just push the notification functionality as-is to production without the workaround measures, and with a warning:

> **Mobile users:** The Shuffle Clock is prone to be put to sleep by your mobile browser or device if not in focus, resulting in dropped alerts (the mobile environment is more aggressive about putting processes to sleep, and a sufficient workaround to this hasn't been found). For best results:
> - Have the Shuffle Clock installed to your homescreen. Instructions for [Chrome](https://support.google.com/chrome/answer/9658361?hl=en&co=GENIE.Platform%3DAndroid) and [Firefox](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Guides/Installing#add_to_home_screen)
> - Keep the Shuffle Clock's page in focus when running, or have it regain focus every shuffle
> - Depending on your device, prevent your device from enacting battery saving optimization or background activity suppression on the browser app you're using

~~I can assist that final point by having Service Worker notifications take the client to the Clock's page (and *then* their desired biome if they have jump enabled)~~ Implemented.

endulum committed:
temporary reversion to using app to invoke notifications

- comment out service worker functionality for scheduling notifications
- create and use SW message for simply invoking a notif right away
- have `onPause` and `onPlay` arguments to `useClock` be optional

vercel will deploy this commit and i will observe the behavior of the
page on my mobile device from the deployment url.

endulum committed:
add current time to test notif text

endulum committed:
suppress vercel deployment

endulum committed:
clean up notifying

- empty sw file, doesn't need event listeners
- recreate notif util file similarly to how it was before this refresh
- - also have notif util function switch notif support type if either
  type fails, allowing for simpler support check without dummy notif
- rename notif util file
- allow app context to export support setter for use in notif util
- remove stray logger

endulum committed:
remove getMils util, bring back getAlertString util

endulum committed:
fix typos in notif state hook

endulum committed:
let notif state hook init notif support type on permission granted

endulum committed:
complete ui for settings

endulum committed:
rename onAlert param  for clock hook, to onTick

endulum committed:
implement proper shuffle alert using alert string

endulum committed:
implement trick to refocus app on notif click

- force notification click to refocus app window
- have sworker post message to client if biome jump is enabled, and have
  event listener on client respond to this message to perform the biome
jump

endulum committed:
temp revert to 5s test notifs, unsuppress vercel deployment

endulum committed:
fix unused variable preventing build?

endulum committed:
revert to proper shuffle timing

endulum committed:
bring in sound functionality

endulum committed:
fix stray import from ignored file

endulum committed:
create error component to display errors

a client who cannot see the console will want to know what kind of error
happened. previously, the clock doesn't react, but this small alert
should help provide a visually captureable hint.

- add error state hook in app context
- use error setter in notification testing, sound testing, shuffle sound
  and notif
- create alert component to display and dismiss error
- listen for alerts from the service worker too
- remove trycatches in notify and playSound functions to allow them to
  propagate to trycatches in notifmgmt and soundmgmt

endulum committed:
fix test buttons playable despite switches disabled

endulum committed:
prevent sw notif handler from closing notif

edenchazard commented on eslint.config.mjs:
```diff
@@ -0,0 +1,51 @@
+        { args: 'all', argsIgnorePattern: '^_' },
+        // circumvent unused var warnings by starting var names with `_`
+      ],
+      'no-console': [
```
so, little trick, there's a setting in the vite config to remove these from build. It's something like

```
esbuild: {
      options: {
        drop: process.env.NODE_ENV === 'production' ? ['console'] : [],
      },
    },
```

endulum replied:
Neat. Did [some research](https://github.com/vitejs/vite/discussions/7920#discussioncomment-4803178) since I wrote and need `error` and `warn` in some places:

```js
esbuild: {
  pure: process.env.NODE_ENV === 'production' ? ['console.log'] : [],
},
```

edenchazard commented on index.html:
```diff
@@ -1,94 +1,40 @@
+    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
+    <link rel="manifest" href="/manifest.json" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Vite + React + TS</title>
```
.

edenchazard commented on src/App.tsx:
```diff
@@ -0,0 +1,96 @@
+        <main>
+          <DelayMgmt />
+          <SoundMgmt />
+          <NotifMgmt />
```
prefer just using the full name, unless it's especially an eyesore.

edenchazard commented on src/AppContext.tsx:
```diff
@@ -0,0 +1,86 @@
+  message: string;
+};
+
+const AppContext = createContext(
+  {} as {
+    // settings
+    clockSettings: IClockSettings;
+    setClockSettings: React.Dispatch<React.SetStateAction<IClockSettings>>;
+    handleInput: <T extends ChangeEvent<HTMLInputElement | HTMLSelectElement>>(
+      event: T
+    ) => void;
+    // notifs
+    permission: TNotifPerms;
+    askPermission: () => Promise<void>;
+    support: TNotifTypes;
+    setSupport: React.Dispatch<React.SetStateAction<TNotifTypes>>;
+    error: Error | null;
+    setError: React.Dispatch<React.SetStateAction<Error | null>>;
+  }
+);
```
```suggestion
const AppContext = createContext<{
    // settings
    clockSettings: IClockSettings;
    setClockSettings: React.Dispatch<React.SetStateAction<IClockSettings>>;
    handleInput: <T extends ChangeEvent<HTMLInputElement | HTMLSelectElement>>(
      event: T
    ) => void;
    // notifs
    permission: TNotifPerms;
    askPermission: () => Promise<void>;
    support: TNotifTypes;
    setSupport: React.Dispatch<React.SetStateAction<TNotifTypes>>;
    error: Error | null;
    setError: React.Dispatch<React.SetStateAction<Error | null>>;
  }>({});
```

this works, right?

endulum replied:
I'm afraid not! `{}` is underlined with:

```
Argument of type '{}' is not assignable to parameter of type '{ clockSettings: IClockSettings; setClockSettings: Dispatch<SetStateAction<IClockSettings>>; handleInput: <T extends ChangeEvent<HTMLInputElement | HTMLSelectElement>>(event: T) => void; ... 5 more ...; setError: Dispatch<...>; }'.
  Type '{}' is missing the following properties from type '{ clockSettings: IClockSettings; setClockSettings: Dispatch<SetStateAction<IClockSettings>>; handleInput: <T extends ChangeEvent<HTMLInputElement | HTMLSelectElement>>(event: T) => void; ... 5 more ...; setError: Dispatch<...>; }': clockSettings, setClockSettings, handleInput, permission, and 5 more.ts(2345)
```

edenchazard replied:
ah because you aren't defining the default values between the `()`

edenchazard commented on src/components/Clock.tsx:
```diff
@@ -0,0 +1,89 @@
+  const { support, setSupport, clockSettings, setError } =
+    useContext(AppContext);
+  const { customAudio } = useCustomAudio();
+  const [isAlerting, setIsAlerting] = useState<boolean>(false);
```
the typehint is redundant here

edenchazard commented on src/components/Clock.tsx:
```diff
@@ -0,0 +1,89 @@
+    onPause: () => {
+      setIsAlerting(false);
+    },
+    onTick: async (time) => {
```
```suggestion
    async onTick(time) {
```

if you like

endulum replied:
Didn't know this was valid syntax, learned something new.

edenchazard commented on src/components/Clock.tsx:
```diff
@@ -0,0 +1,89 @@
+
+  useDocumentTitle(
+    isPaused
+      ? 'Cave Shuffle Clock'
```
ah... I see now. I'd still set the document title tag anyway just for seo purposes.

edenchazard commented on src/components/Clock.tsx:
```diff
@@ -0,0 +1,89 @@
+      <button
+        type="button"
+        className={`clock-button ${
+          isAlerting ? 'alerting' : isPaused ? 'pausing' : 'playing'
```
try to avoid nested ternary? they can be a real plague to read

endulum replied:
How about...
```jsx
className={`clock-button ${
  (() => {
    if (isAlerting) return 'alerting';
    if (isPaused) return 'pausing';
    return 'playing';
  })()
} p-05`}
```

edenchazard commented on src/components/Clock.tsx:
```diff
@@ -0,0 +1,89 @@
+            playSound(clockSettings, customAudio);
+          } catch (e) {
+            console.error(e);
+            setError({
+              type: 'Error playing shuffle sound',
+              message:
+                e instanceof Error ? e.message : 'See console for details.',
+            });
```
does this throw up a toast? will the user actually be notified of the problem?

endulum replied:
The `Error` component basically "listens" for errors using `setError` and will display a plain dismissable alert dialog under the clock.

![2025-03-23-141711_hyprshot](https://github.com/user-attachments/assets/c9183c97-9634-471b-a64e-11577e230ee3)

edenchazard commented on src/components/DelayMgmt.tsx:
```diff
@@ -0,0 +1,85 @@
+  const { clockSettings, handleInput } = useContext(AppContext);
+  return (
+    <div className="flex-col">
+      <label htmlFor="delay" className="flex-row jcspb w100">
+        <b>Delay:</b>
+        <span>
+          <input
+            id="delay"
+            type="number"
+            min="1"
+            max="59"
+            title="delay in seconds"
+            defaultValue={clockSettings.delay}
+            onChange={handleInput}
+          />
+          &nbsp; seconds before shuffle
+        </span>
+      </label>
```
not sure what the advice is re. non-input elements nested inside labels. you'll need to research to confirm, but I feel it may be problematic.

endulum replied:
No worries, I can suffice for

```jsx
<div className="flex-row jcspb w100">
  <label htmlFor="delay">
    <b>Delay:</b>
  </label>
  <span>
    <input
      id="delay"
      type="number"
      min="1"
      max="59"
      title="delay in seconds"
      defaultValue={clockSettings.delay}
      onChange={handleInput}
    />
    &nbsp; seconds before shuffle
  </span>
</div>
```

edenchazard commented on src/components/Error.tsx:
```diff
@@ -0,0 +1,50 @@
+  if (error)
+    return (
+      <div className="error flex-row ais g-05 p-05 w100">
+        <button
```
don't forget your `type=button`!

edenchazard commented on src/components/Error.tsx:
```diff
@@ -0,0 +1,50 @@
+    if ('serviceWorker' in navigator && navigator.serviceWorker.controller) {
+      navigator.serviceWorker.addEventListener('message', (event): void => {
+        if (
+          event.data &&
+          event.data.type === 'error' &&
+          'error' in event.data
```
```suggestion
          event.data?.type === 'error' &&  event.data?.error
```

optional chaining syntax works quite nicely here

edenchazard commented on src/components/Footer.tsx:
```diff
@@ -0,0 +1,37 @@
+      </div>
+      <p>
+        Raise your catches at the{' '}
+        <a href="https://chazza.me/dc/hatchery">Garden of Eden!</a>
```
YOOOOOOOOOOOO LOOOOK THAT MY SITE!!!!!!!!!!!!

edenchazard commented on src/AppContext.tsx:
```diff
@@ -0,0 +1,86 @@
+import { useState, createContext, type ChangeEvent } from 'react';
+
+import { IClockSettings, TNotifPerms, TNotifTypes } from './types';
```
if these are just types, use `import type { ... }` or ` import { type A, B, type C }`. this will reduce your bundle sizes just a tad

edenchazard commented on src/components/SoundMgmt.tsx:
```diff
@@ -0,0 +1,153 @@
+          )}
+          <label htmlFor="soundUpload" className="row">
+            <input
+              aria-label="upload custom sound as a file"
```
it's nicer to have a label with a `sr-only` class or an `aria-labelledby` attribute. Reason being, this `aria-label` won't be picked up by browser translation services.

endulum replied:
This exposed me to the fact I forgot to take out the `row` classes - `.row` doesn't exist in this project. Goes to show how lazy I got with copypasting.

That being said, does this work?

```jsx
<label aria-labelledby="soundUpload">
  <input
    aria-label="upload custom sound as a file"
    // ...
```

edenchazard commented on src/functions/getAudioString.ts:
```diff
@@ -0,0 +1,19 @@
+      const audioString: string = JSON.stringify(e.target?.result);
+      const { size } = new Blob([audioString]);
+      if (size > 2500000) {
+        alert(
+          'Uploaded sound is too large. Only sounds under 2MB can be stored.\n'
+        );
+        res(null);
+      } else res(audioString);
```
this a nicer job for your setError?

edenchazard commented on src/functions/transformSettings.ts:
```diff
@@ -0,0 +1,28 @@
+export default function transformSettings(data: unknown): IClockSettings {
+  if (data === null || typeof data !== 'object') return defaults;
+
+  const transformedSettings: keyable = {};
+
+  Object.keys(defaults).forEach((setting) => {
+    if (
+      // setting is not defined in data
+      (data as keyable)[setting] === undefined ||
+      // setting is the wrong type in data
+      typeof (data as keyable)[setting] !==
+        typeof defaults[setting as keyof IClockSettings]
+    ) {
+      // use the default value of this setting instead of data's value
+      transformedSettings[setting] = defaults[setting as keyof IClockSettings];
+    } else {
+      // if everything's ok, use data's value for this setting
+      transformedSettings[setting] = (data as keyable)[setting];
+    }
+  });
```
Why not set `transformedSettings` to be a clone of the default settings, and then as you iterate and check things, you can replace the value? then you get to drop an else statement.

edenchazard reviewed (commented):
nice codes bro

endulum committed:
address comments

- adjust vite config to drop console logs on build
- give index.html proper page title
- lengthen "mgmt" to "management" in component names
- remove redundant boolean state typehints
- use iife instead of nested ternary in `Clock` component
- un-nest span from label in `DelayManagement` component
- fix forgotten `type="button"` attributes
- add operational chaining in `Error` component
- fix type import syntax
- try `aria-labelledby` in `Footer` component
- rewrite `handleAudioInput` and `getAudioString` to allow `setError` to
  reflect errors thrown when uploading sounds
- simplify `transformSettings` bu cloning default settings and only
  replacing values

endulum committed:
remove stray nonexistent class

endulum committed:
turn sound grid into a real grid

endulum committed:
spruce up readme

edenchazard commented on src/assets/styles/main.css:
```diff
@@ -84,6 +84,15 @@ select {
 }
+label {
+  white-space: nowrap;
+  &.sr-only {
+    position: absolute;
+    width: 1px;
+    height: 1px;
+    overflow: hidden;
+  }
```
unscope this from the label. `sr-only` is a utility class and not generally purposely tied to label. you're also missing properties like `clip`

edenchazard commented on src/components/SoundManagement.tsx:
```diff
@@ -86,72 +86,60 @@ export function SoundManagement() {
-      </div>
+          {new Array(5)
+            .fill(0)
+            .map((_dummy, index) => `SMS Alert ${index + 1}`)
```
Suggest using `_` for unused variables in line with general expectations.

edenchazard commented on README.md:
```diff
@@ -1,34 +1,48 @@
 ### Troubleshooting
 
-* __Time is inaccurate:__ The tool uses your local time. Be sure that your device clocked is accurately synced.
+**Mobile users:** The Shuffle Clock is prone to be put to sleep by your mobile browser or device if not in focus, resulting in dropped alerts. The mobile environment is more aggressive about putting processes to sleep, and "forcing" the page to bypass mobile sleep is out of my control. If you're experiencing dropped alerts, ensure that you:
```
```suggestion
**Mobile users:** Mobile browsers tend to sleep inactive pages when not in focus, resulting in dropped alerts. To minimise this, ensure that you:
```

endulum committed:
shorten mobile user warning on readme

Co-authored-by: edenchazard <38665888+edenchazard@users.noreply.github.com>

endulum committed:
address comments

- use `_` for totally unused variable
- unscope and improve `.sr-only`

endulum committed:
create faq modal and move troubleshooting info from readme to faq

endulum committed:
quick fix for crashing on ios

endulum committed:
add meta tags for link preview

endulum committed:
tweak bold

endulum committed:
Merge branch 'main' into 2025-remake
